Mathlib/Algebra/Category/ModuleCat/Presheaf/Generator.lean

```diff
@@ -0,0 +1,206 @@
+/-
+Copyright (c) 2024 Joël Riou. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joël Riou
+-/
+import Mathlib.Algebra.Category.ModuleCat.Presheaf.Abelian
+import Mathlib.Algebra.Category.ModuleCat.Presheaf.EpiMono
+import Mathlib.Algebra.Category.ModuleCat.Presheaf.Free
+import Mathlib.Algebra.Homology.ShortComplex.Exact
+import Mathlib.CategoryTheory.Elements
+import Mathlib.CategoryTheory.Generator
+
+/-!
+# Generators for the category of presheaves of modules
+
+In this file, given a presheaf of rings `R` on a category `C`,
+we construct a set `FreeYoneda R` of presheaves of modules that
+is separating and detecting: this set consists of presheaves of
+modules of form `(free R).obj (yoneda.obj X)` for `X : C`,
+i.e. free presheaves of modules generated by the Yoneda
+presheaf represented by some `X : C` (the functor
+represented by such a presheaf of modules is the evaluation
+functor `M ↦ M.obj (op X)`, see `freeYonedaEquiv`).
+
+We deduce that if `C : Type u` is a small category,
+and `R : Cᵒᵖ ⥤ RingCat.{u}`, then `PresheafOfModules.{u} R`
+is a well powered category.
+
+-/
+
+universe v v₁ u u₁
+
+open CategoryTheory Limits
+
+namespace PresheafOfModules
+
+variable {C : Type u} [Category.{v} C] {R : Cᵒᵖ ⥤ RingCat.{v}}
+  {C₀ : Type u} [SmallCategory C₀] {R₀ : C₀ᵒᵖ ⥤ RingCat.{u}}
+
+/-- When `R : Cᵒᵖ ⥤ RingCat`, `M : PresheafOfModules R`, and `X : C`, this is the
+bijection `((free R).obj (yoneda.obj X) ⟶ M) ≃ M.obj (Opposite.op X)`. -/
+noncomputable def freeYonedaEquiv {M : PresheafOfModules.{v} R} {X : C} :
+    ((free R).obj (yoneda.obj X) ⟶ M) ≃ M.obj (Opposite.op X) :=
+  freeHomEquiv.trans yonedaEquiv
+
+lemma freeYonedaEquiv_symm_apply (M : PresheafOfModules.{v} R) (X : C)
+    (x : M.obj (Opposite.op X)) :
+    (freeYonedaEquiv.symm x).app (Opposite.op X) (ModuleCat.freeMk (𝟙 _)) = x := by
+  dsimp [freeYonedaEquiv, freeHomEquiv, yonedaEquiv]
+  rw [ModuleCat.freeDesc_apply, op_id, M.presheaf.map_id]
+  rfl
+
+lemma freeYonedaEquiv_comp {M N : PresheafOfModules.{v} R} {X : C}
+    (m : ((free R).obj (yoneda.obj X) ⟶ M)) (φ : M ⟶ N) :
+    freeYonedaEquiv (m ≫ φ) = φ.app _ (freeYonedaEquiv m) := rfl
+
+variable (R) in
+/-- The set of `PresheafOfModules.{v} R` consisting of objects of the
+form `(free R).obj (yoneda.obj X)` for some `X`.  -/
+inductive FreeYoneda : Set (PresheafOfModules.{v} R)
+  | free_yoneda_mem (X : C) : FreeYoneda ((free R).obj (yoneda.obj X))
```

Comment:
This sees to be abusing the `Set = _ -> Prop` defeq. Either replace it with `{ (free R).obj (yoneda.obj X) | (X : C) }` (if it works; not tested) and renamed to `freeYoneda`, or define it as `_ -> Prop` and use `setOf` below.

Reply:
Thanks for the suggestion. I have just fixed this.